lib/editor/actions/map/stopStrategies.js

```diff
@@ -220,26 +220,42 @@ export function addStopAtInterval (latlng: LatLng, activePattern: Pattern, contr
   }
 }
 
-export function addStopToPattern (pattern: Pattern, stop: GtfsStop, index?: ?number) {
+export function addStopToPattern (pattern: Pattern, stop: GtfsStop | GtfsLocation, index?: ?number) {
+  // eslint-disable-next-line complexity
   return async function (dispatch: dispatchFn, getState: getStateFn) {
     const {data, editSettings} = getState().editor
     const {followStreets} = editSettings.present
-    const {patternStops: currentPatternStops, shapePoints} = pattern
+    const {
+      patternLocations: currentPatternLocations,
+      patternStops: currentPatternStops,
+      shapePoints
+    } = pattern
+    const patternLocations = clone(currentPatternLocations)
     const patternStops = clone(currentPatternStops)
     const {controlPoints, patternSegments} = getControlPoints(getState())
     const patternLine = lineString(coordinatesFromShapePoints(shapePoints))
     const hasShapePoints = shapePoints && shapePoints.length > 2
+    let patternHalts = mergePatternHalts(patternStops, patternLocations)
     const newStop = stopToPatternStop(
       stop,
       (typeof index === 'undefined' || index === null)
-        ? patternStops.length
+        ? patternHalts.length
         : index
     )
 
-    if (typeof index === 'undefined' || index === null || index === patternStops.length) {
-      // Push pattern stop to cloned list.
-      patternStops.push(newStop)
-      if (hasShapePoints) {
+    // $FlowFixMe Flow doesn't understand type check
```

Comment:
Can some of the advice in https://flow.org/en/docs/lang/width-subtyping/ help remove these exceptions to flow? (not sure if they work with the version we are using though)

Reply:
Yeah I tried doing this and only got really obscure errors. I think we'd have to upgrade as a separate PR